src/extras.jl

```diff
@@ -0,0 +1,143 @@
+function fill_refs!(refs::AbstractVector, X::AbstractArray,
+                    breaks::AbstractVector, extend::Bool, nullok::Bool)
+    n = length(breaks)
+    lower = first(breaks)
+    upper = last(breaks)
+
+    @inbounds for i in eachindex(X)
+        x = X[i]
+
+        if extend && x == upper
+            refs[i] = n-1
+        elseif !extend && !(lower <= x < upper)
+            throw(ArgumentError("value $x (at index $i) is outside the breaks: adapt them manually, or pass extend=true"))
```

Comment:
The `extend=true` part is a bit misleading since `extend` isn't a keyword argument, actually passing `extend=true` will be an error.